src/main/scala/mesosphere/marathon/core/storage/zookeeper/ZooKeeperPersistenceStore.scala

```diff
@@ -0,0 +1,274 @@
+    *
+    * @return
+    */
+  override def deleteFlow: Flow[String, String, NotUsed] =
+    Flow[String]
+      .via(logPath("Deleting a node at "))
+      .via(metric(deleteMetric))
+      .mapAsync(parallelism)(path =>
+        factory
+          .delete()
+          .forPath(path)
+          .toScala
+          .map(_ => path)
+      )
+
+  override def delete(path: String): Future[String] = {
+    logger.debug(s"Deleting a node at $path")
+    deleteMetric.increment()
+    factory
+      .delete()
+      .forPath(path)
+      .toScala
+      .map(_ => path)
+  }
+
+  /**
+    * A Flow returning a stream of children for the given parent nodes.
+    *
+    * @return
+    */
+  override def childrenFlow: Flow[String, Seq[String], NotUsed] =
+    Flow[String]
+      .via(logPath("Getting children at "))
+      .via(metric(childrenMetric))
+      .mapAsync(parallelism)(path =>
+        factory
+          .children()
+          .forPath(path)
+          .toScala
+          .map(JavaConverters.asScalaBuffer(_).to[Seq])
+          .map(children => children.map(child => s"$path/$child")) // Return full path
+      )
+
+  override def children(path: String): Future[Seq[String]] = {
+    childrenMetric.increment()
+    logger.debug(s"Getting children at $path")
+    factory
+      .children()
+      .forPath(path).toScala
+      .map(JavaConverters.asScalaBuffer(_).to[Seq])
+      .map(children => children.map(child => s"$path/$child")) // Return full path
+  }
+
+  override def existsFlow: Flow[String, Boolean, NotUsed] =
+    Flow[String]
+      .via(metric(existsMetric))
+      .via(logPath("Checking node existence for "))
+      .mapAsync(parallelism)(path =>
+        factory
+          .checkExists()
+          .forPath(path).toScala
+          .map(stat => if (stat == null) false else true)
+      )
+
+  override def exists(path: String): Future[Boolean] = {
+    existsMetric.increment()
+    logger.debug(s"Checking node existence for $path")
+    factory
+      .checkExists()
+      .forPath(path).toScala
+      .map(stat => if (stat == null) false else true)
+  }
+
+  override def sync(path: String = "/"): Future[Done] = {
+    logger.debug(s"Syncing nodes for path $path")
+    factory
+      .sync()
+      .forPath(path).toScala
+      .map(_ => Done)
+  }
+
+  /**
+    * Method takes a list of transaction [[mesosphere.marathon.core.storage.zookeeper.PersistenceStore.StoreOp]] operations
+    * and submits them. An exception is thrown if one of the operations fail. Currently only create, update, delete and
+    * check operations are supported.
+    *
+    * Note: Due to current state of the underlying Curator API, [[mesosphere.marathon.core.storage.zookeeper.PersistenceStore.CreateOp]]s
+    * can't create parent nodes for nested paths if parent nodes does not exist.
+    *
+    * @param operations a list of transaction operations
+    * @return
+    */
+  override def transaction(operations: Seq[StoreOp]): Future[Done] = {
```

Comment:
How about a transaction builder instead of the ops?

```scala
val result: Future[Done] = transaction().create(node).read(node.path).delete(node.path).run()
```

Reply:
I'll leave it open for the subsequent PR, maybe even from you ;)